File: app/src/main/java/app/oswaldogh/plazashop/Ui/Products/ProductsPresenter.java
package app.oswaldogh.plazashop.Ui.Products;

import java.util.ArrayList;

import app.oswaldogh.plazashop.Entities.Product;

/**
 * Created by oswaldogh89 on 04/01/18.
 */

public final class ProductsPresenter implements Interface.Presenter {
    private Interface.Model model;
    private Interface.View view;

    ProductsPresenter(Interface.View view) {
        this.view = view;
        this.model = new ProductsModel(this);
    }

    void getDataProducts(int totalProducts) {
        if (this.view != null) {
            this.model.getProducts(totalProducts);
        }
    }

    public void onLoadProucts(ArrayList<Product> products) {
        if (this.view != null) {
            if (this.view.adapterIsNull()) {
                this.view.createListProducts(products);
            } else {
                if (products.size() > 0)
                    this.view.addMoreProductsToList(products);
            }
            if (products.size() == 0 && this.view.adapterIsNull()) {
                this.view.setProductListEmpty();
            }
            this.view.hideLoadMoreProgress();
        }
    }
}

File: app/src/main/java/app/oswaldogh/plazashop/Ui/Products/ProductsPresenter.kt
package app.oswaldogh.plazashop.Ui.Products

import app.oswaldogh.plazashop.Entities.Product
import java.util.*

/**
 * Created by oswaldogomez on 05/12/17.
 */

class ProductsPresenter internal constructor(private val view: Interface.View?) : Interface.Presenter {
    private val model: Interface.Model

    init {
        this.model = ProductsModel(this)
    }

    fun getDataProducts(totalProducts: Int) {
        if (view != null) {
            model.getProducts(totalProducts)
        }
    }

    override fun onLoadProucts(products: ArrayList<Product>) {
        view?.showProducts(products)
    }
}
